core/src/main/java/com/linecorp/armeria/common/ByteArrayHttpData.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+
+import com.google.common.base.MoreObjects;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.buffer.Unpooled;
+import io.netty.util.internal.EmptyArrays;
+import it.unimi.dsi.fastutil.io.FastByteArrayInputStream;
+
+/**
+ * A {@code byte[]}-based {@link HttpData}.
+ */
+final class ByteArrayHttpData implements HttpData {
+
+    static final ByteArrayHttpData EMPTY = new ByteArrayHttpData(EmptyArrays.EMPTY_BYTES, false);
+    static final ByteArrayHttpData EMPTY_EOS = new ByteArrayHttpData(EmptyArrays.EMPTY_BYTES, true);
+
+    private final byte[] array;
+    private final boolean endOfStream;
+
+    ByteArrayHttpData(byte[] array) {
+        this(array, false);
+    }
+
+    private ByteArrayHttpData(byte[] array, boolean endOfStream) {
+        this.array = array;
+        this.endOfStream = endOfStream;
+    }
+
+    @Override
+    public byte[] array() {
+        return array;
+    }
+
+    @Override
+    public int length() {
+        return array.length;
+    }
+
+    @Override
+    public String toString(Charset charset) {
+        requireNonNull(charset, "charset");
+        return new String(array, charset);
+    }
+
+    @Override
+    public String toString() {
+        final MoreObjects.ToStringHelper helper =
+                MoreObjects.toStringHelper(this)
+                           .omitNullValues()
+                           .addValue(length() + "B");
+
+        if (isEndOfStream()) {
+            helper.addValue("endOfStream");
+        }
+        return helper.add("array", array)
+                     .toString();
+    }
+
+    @Override
+    public InputStream toInputStream() {
+        return new FastByteArrayInputStream(array);
+    }
+
+    @Override
+    public boolean isEndOfStream() {
+        return endOfStream;
+    }
+
+    @Override
+    public ByteArrayHttpData withEndOfStream(boolean endOfStream) {
+        if (isEmpty()) {
+            return endOfStream ? EMPTY_EOS : EMPTY;
+        }
+
+        if (isEndOfStream() == endOfStream) {
```

Comment:
I think this should come first for readability given the method's primary goal is changing the value of endofstream